In [1]:
# Import required python libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Reading Leads dataset
leads = pd.read_csv('Leads.csv')

# printing the dataset .head()-print 1st five rows of dataset
leads.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  ...  \
0                            0                   0.0  ...   
1                          674                   2.5  ...   
2                         1532                   2.0  ...   
3                          305                   1.0  ...   
4                         1428                   1.0  ...   

  Get updates on DM Content    Lead Profile    City  \
0                        No          Select  Select   
1                        No          Select  Select   
2                        No  Potential Lead  Mumbai   
3                        No          Select  Mumbai   
4                        No          Select  Mumbai   

  Asymmetrique Activity Index Asymmetrique Profile Index  \
0                   02.Medium                  02.Medium   
1                   02.Medium                  02.Medium   
2                   02.Medium                    01.High   
3                   02.Medium                    01.High   
4                   02.Medium                    01.High   

  Asymmetrique Activity Score Asymmetrique Profile Score  \
0                        15.0                       15.0   
1                        15.0                       15.0   
2                        14.0                       20.0   
3                        13.0                       17.0   
4                        15.0                       18.0   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   
3                                       No   
4                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified  

[5 rows x 37 columns]

In [3]:
# Shape of the dataset contains 9240 rows and 37 column
leads.shape

(9240, 37)

In [4]:
# Printing cloumn names of 37 columns
leads.columns

Index(['Prospect ID', 'Lead Number', 'Lead Origin', 'Lead Source',
       'Do Not Email', 'Do Not Call', 'Converted', 'TotalVisits',
       'Total Time Spent on Website', 'Page Views Per Visit', 'Last Activity',
       'Country', 'Specialization', 'How did you hear about X Education',
       'What is your current occupation',
       'What matters most to you in choosing a course', 'Search', 'Magazine',
       'Newspaper Article', 'X Education Forums', 'Newspaper',
       'Digital Advertisement', 'Through Recommendations',
       'Receive More Updates About Our Courses', 'Tags', 'Lead Quality',
       'Update me on Supply Chain Content', 'Get updates on DM Content',
       'Lead Profile', 'City', 'Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score',
       'I agree to pay the amount through cheque',
       'A free copy of Mastering The Interview', 'Last Notable Activity'],
      dtype='object')

In [5]:
# Checking the info to see the types of the values and the null values if any
leads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
Prospect ID                                      9240 non-null object
Lead Number                                      9240 non-null int64
Lead Origin                                      9240 non-null object
Lead Source                                      9204 non-null object
Do Not Email                                     9240 non-null object
Do Not Call                                      9240 non-null object
Converted                                        9240 non-null int64
TotalVisits                                      9103 non-null float64
Total Time Spent on Website                      9240 non-null int64
Page Views Per Visit                             9103 non-null float64
Last Activity                                    9137 non-null object
Country                                          6779 non-null object
Specialization                                   7802 

Data Cleaning and Preparation

In [6]:
# Checking missing values count in each column
leads.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                          2461
Specialization                                   1438
How did you hear about X Education               2207
What is your current occupation                  2690
What matters most to you in choosing a course    2709
Search                                              0
Magazine                                            0
Newspaper Article           

From above results its clear that columns with high numbers of null values are not useful so lets drop columns whose null value count is more than 3000 and try it.

In [7]:
# Dropping columns with 3000+ null values

for column in leads.columns:
    if leads[column].isnull().sum() > 3000:
        leads.drop(column, 1, inplace=True)

In [8]:
# Checking missing values count in each column again
leads.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                          2461
Specialization                                   1438
How did you hear about X Education               2207
What is your current occupation                  2690
What matters most to you in choosing a course    2709
Search                                              0
Magazine                                            0
Newspaper Article           

In [9]:
leads.drop(['City','Country'], axis = 1, inplace = True)

In [10]:
# leads.drop(['Country'], axis = 1, inplace = True)

In [11]:
# Checking percentage of missing values in each column

round(100*(leads.isnull().sum()/len(leads.index)), 2)

Prospect ID                                       0.00
Lead Number                                       0.00
Lead Origin                                       0.00
Lead Source                                       0.39
Do Not Email                                      0.00
Do Not Call                                       0.00
Converted                                         0.00
TotalVisits                                       1.48
Total Time Spent on Website                       0.00
Page Views Per Visit                              1.48
Last Activity                                     1.11
Specialization                                   15.56
How did you hear about X Education               23.89
What is your current occupation                  29.11
What matters most to you in choosing a course    29.32
Search                                            0.00
Magazine                                          0.00
Newspaper Article                                 0.00
X Educatio

In [12]:
# Value counts of all the columns

for column in leads:
    print(leads[column].astype('category').value_counts())
    print('___________________________________________________')

fffb0e5e-9f92-4017-9f42-781a69da4154    1
56453aec-3f7b-4f30-870c-8f966d393100    1
53ac14bd-2bb2-4315-a21c-94562d1b6b2d    1
53aabd84-5dcc-4299-bbe3-62f3764b07b1    1
539ffa32-1be7-4fe1-b04c-faf1bab763cf    1
539eb309-df36-4a89-ac58-6d3651393910    1
5398e7ff-74db-4074-89fb-4fd9a603f521    1
53953744-234a-4cb9-9af4-bcc47eb472f4    1
539366d9-f633-455a-99e4-dbc5907db28e    1
5390c5fe-b12c-4f6e-ae92-908672abb0a1    1
5379ee79-64b7-44f8-8c56-0e1ca2d5b887    1
537963c8-22d9-459d-8aae-ddac40580ffb    1
53744d5a-0483-42c0-80b0-8990a4d2356d    1
53715ab1-2106-4c4e-8493-81cc465eb9ce    1
536cdc6b-f4c1-449d-bfd8-9ef0ac912dbb    1
53690d88-52f0-4ce5-b6b8-a13570a6db35    1
5363bd79-576c-48ed-83e4-024c81ea00c5    1
53c4e210-3344-4737-813f-74ef9a747ab6    1
53dbb914-71e7-458a-9749-cfb4d655eac2    1
53dd16bd-8201-448d-8e20-97de1cf44a7f    1
541325bd-15bb-4b52-8ad9-3fdf3cb1dd55    1
5434ccf3-9de6-4c72-8dd6-66c2829d0ee2    1
542a0891-2e52-40ba-ab42-e468b9636322    1
54238b21-65ce-4304-98c6-0f8a6b9671

These 3 columns (Lead Profile,How did you hear about X Education & Specialization) have the level 'Select'. And in most of the cases students choosed 'Select' which basically means that the student had not selected the option for that particular column. These values can be treated as missing values. Lets count the number of "Select' values in these 3 columns.

In [13]:
leads['Lead Profile'].astype('category').value_counts()

Select                         4146
Potential Lead                 1613
Other Leads                     487
Student of SomeSchool           241
Lateral Student                  24
Dual Specialization Student      20
Name: Lead Profile, dtype: int64

In [14]:
leads['How did you hear about X Education'].value_counts()

Select                   5043
Online Search             808
Word Of Mouth             348
Student of SomeSchool     310
Other                     186
Multiple Sources          152
Advertisements             70
Social Media               67
Email                      26
SMS                        23
Name: How did you hear about X Education, dtype: int64

In [15]:
leads['Specialization'].value_counts()

Select                               1942
Finance Management                    976
Human Resource Management             848
Marketing Management                  838
Operations Management                 503
Business Administration               403
IT Projects Management                366
Supply Chain Management               349
Banking, Investment And Insurance     338
Media and Advertising                 203
Travel and Tourism                    203
International Business                178
Healthcare Management                 159
Hospitality Management                114
E-COMMERCE                            112
Retail Management                     100
Rural and Agribusiness                 73
E-Business                             57
Services Excellence                    40
Name: Specialization, dtype: int64

Lead Profile and How did you hear about X Education Select value count is more than 70% of total value so lets drop these 2 column. 

And lets drop the columns where the value is NO, Columns are - Do Not Call, Search, Magazine, Newspaper Article, X Education Forums, Newspaper, Digital Advertisement, Through Recommendations, Receive More Updates About Our Courses, Update me on Supply Chain Content, Get updates on DM Content, I agree to pay the amount through cheque.

Also, lets drop the column "What matters most to you in choosing a course" as out of 3 catagories - catagory "Better Career Prospects was selected 6528 times while the other two catagories only 1 time and 2 times. 

Lets drop all these above mentioned columns as it will not help in analysis.

In [16]:
leads.drop(['Do Not Call', 'Search', 'Magazine', 'Newspaper Article', 'X Education Forums', 'Newspaper', 
            'Digital Advertisement', 'Through Recommendations', 'Receive More Updates About Our Courses', 
            'Update me on Supply Chain Content', 'Get updates on DM Content', 
            'I agree to pay the amount through cheque','What matters most to you in choosing a course','Lead Profile', 'How did you hear about X Education'], axis = 1, inplace = True)

In [17]:
# Checking missing values count in each column again
leads.isnull().sum()

Prospect ID                                  0
Lead Number                                  0
Lead Origin                                  0
Lead Source                                 36
Do Not Email                                 0
Converted                                    0
TotalVisits                                137
Total Time Spent on Website                  0
Page Views Per Visit                       137
Last Activity                              103
Specialization                            1438
What is your current occupation           2690
A free copy of Mastering The Interview       0
Last Notable Activity                        0
dtype: int64

In [18]:
# Let's drop the null rows for the column 'What is you current occupation'
leads = leads[~pd.isnull(leads['What is your current occupation'])]

In [19]:
# Checking missing values count in each column again
leads.isnull().sum()

Prospect ID                                 0
Lead Number                                 0
Lead Origin                                 0
Lead Source                                36
Do Not Email                                0
Converted                                   0
TotalVisits                               130
Total Time Spent on Website                 0
Page Views Per Visit                      130
Last Activity                             103
Specialization                             18
What is your current occupation             0
A free copy of Mastering The Interview      0
Last Notable Activity                       0
dtype: int64

Now we have a dataset with very less count of null values and we saw a decrease in null values count after removing the null value rows of column 'What is you current occupation'. Now lets remove all null value rows from columns 'Lead Source', 'Total Visits','Page Views Per Visit','Last Activity','Specialization' 1 by 1, start with higher value count i.e. Total visits 130

In [20]:
leads = leads[~pd.isnull(leads['TotalVisits'])]

In [21]:
leads.isnull().sum()

Prospect ID                                0
Lead Number                                0
Lead Origin                                0
Lead Source                               29
Do Not Email                               0
Converted                                  0
TotalVisits                                0
Total Time Spent on Website                0
Page Views Per Visit                       0
Last Activity                              0
Specialization                            18
What is your current occupation            0
A free copy of Mastering The Interview     0
Last Notable Activity                      0
dtype: int64

In [22]:
leads = leads[~pd.isnull(leads['Lead Source'])]
leads.isnull().sum()

Prospect ID                                0
Lead Number                                0
Lead Origin                                0
Lead Source                                0
Do Not Email                               0
Converted                                  0
TotalVisits                                0
Total Time Spent on Website                0
Page Views Per Visit                       0
Last Activity                              0
Specialization                            18
What is your current occupation            0
A free copy of Mastering The Interview     0
Last Notable Activity                      0
dtype: int64

In [23]:
leads = leads[~pd.isnull(leads['Specialization'])]
leads.isnull().sum()

Prospect ID                               0
Lead Number                               0
Lead Origin                               0
Lead Source                               0
Do Not Email                              0
Converted                                 0
TotalVisits                               0
Total Time Spent on Website               0
Page Views Per Visit                      0
Last Activity                             0
Specialization                            0
What is your current occupation           0
A free copy of Mastering The Interview    0
Last Notable Activity                     0
dtype: int64

Atlast we have a dataset with zero NULL Values. Lets proceed with further analysis.

In [24]:
# Checking quality of current dataset after dropping all irrelevant columns and rows
print(len(leads.index))
print(len(leads.index)/9240)                            

6373
0.6897186147186147


In [25]:
# Current dataset after modification
leads.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email  Converted  TotalVisits  \
0      Olark Chat           No          0          0.0   
1  Organic Search           No          0          5.0   
2  Direct Traffic           No          1          2.0   
3  Direct Traffic           No          0          1.0   
4          Google           No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit            Last Activity  \
0                            0                   0.0  Page Visited on Website   
1                          674                   2.5             Email Opened   
2                         1532                   2.0             Email Opened   
3                          305                   1.0              Unreachable   
4                         1428                   1.0        Converted to Lead   

            Specialization What is your current occupation  \
0                   Select                      Unemployed   
1                   Select                      Unemployed   
2  Business Administration                         Student   
3    Media and Advertising                      Unemployed   
4                   Select                      Unemployed   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified

In [26]:
leads.drop(['Prospect ID', 'Lead Number'], 1, inplace = True)
leads.head()

Lead Origin     Lead Source Do Not Email  Converted  \
0                      API      Olark Chat           No          0   
1                      API  Organic Search           No          0   
2  Landing Page Submission  Direct Traffic           No          1   
3  Landing Page Submission  Direct Traffic           No          0   
4  Landing Page Submission          Google           No          1   

   TotalVisits  Total Time Spent on Website  Page Views Per Visit  \
0          0.0                            0                   0.0   
1          5.0                          674                   2.5   
2          2.0                         1532                   2.0   
3          1.0                          305                   1.0   
4          2.0                         1428                   1.0   

             Last Activity           Specialization  \
0  Page Visited on Website                   Select   
1             Email Opened                   Select   
2             Email Opened  Business Administration   
3              Unreachable    Media and Advertising   
4        Converted to Lead                   Select   

  What is your current occupation A free copy of Mastering The Interview  \
0                      Unemployed                                     No   
1                      Unemployed                                     No   
2                         Student                                    Yes   
3                      Unemployed                                     No   
4                      Unemployed                                     No   

  Last Notable Activity  
0              Modified  
1          Email Opened  
2          Email Opened  
3              Modified  
4              Modified

Lets create Dummy Variables for categorical variables. Lets identify the categorical variables first.

In [27]:
# Checking the columns type 'object'

obj = leads.loc[:, leads.dtypes == 'object']
obj.columns

Index(['Lead Origin', 'Lead Source', 'Do Not Email', 'Last Activity',
       'Specialization', 'What is your current occupation',
       'A free copy of Mastering The Interview', 'Last Notable Activity'],
      dtype='object')

In [28]:
# Creating dummy variables using 'get_dummies'
dummy_var = pd.get_dummies(leads[['Lead Origin', 'Lead Source', 'Do Not Email', 'Last Activity',
                              'What is your current occupation','A free copy of Mastering The Interview', 
                              'Last Notable Activity']], drop_first=True)

# Concatinate the dummy_var results to the master dataframe
leads = pd.concat([leads, dummy_var], axis=1)

In [29]:
# Creating dummy variable for the variable 'Specialization' as it got a level 'Select' so lets drop that level by specifying it explicitly

dummy_spl = pd.get_dummies(leads['Specialization'], prefix = 'Specialization')
dummy_spl = dummy_spl.drop(['Specialization_Select'], 1)
leads = pd.concat([leads, dummy_spl], axis = 1)

In [30]:
# Dropping the variables for which dummy variables are created

leads = leads.drop(['Lead Origin', 'Lead Source', 'Do Not Email', 'Last Activity',
                   'Specialization', 'What is your current occupation',
                   'A free copy of Mastering The Interview', 'Last Notable Activity'], 1)

In [31]:
# New updated dataset
leads.head()

Converted  TotalVisits  Total Time Spent on Website  Page Views Per Visit  \
0          0          0.0                            0                   0.0   
1          0          5.0                          674                   2.5   
2          1          2.0                         1532                   2.0   
3          0          1.0                          305                   1.0   
4          1          2.0                         1428                   1.0   

   Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
0                                    0                          0   
1                                    0                          0   
2                                    1                          0   
3                                    1                          0   
4                                    1                          0   

   Lead Origin_Lead Import  Lead Source_Direct Traffic  Lead Source_Facebook  \
0                        0                           0                     0   
1                        0                           0                     0   
2                        0                           1                     0   
3                        0                           1                     0   
4                        0                           0                     0   

   Lead Source_Google  ...  Specialization_IT Projects Management  \
0                   0  ...                                      0   
1                   0  ...                                      0   
2                   0  ...                                      0   
3                   0  ...                                      0   
4                   1  ...                                      0   

   Specialization_International Business  Specialization_Marketing Management  \
0                                      0                                    0   
1                                      0                                    0   
2                                      0                                    0   
3                                      0                                    0   
4                                      0                                    0   

   Specialization_Media and Advertising  Specialization_Operations Management  \
0                                     0                                     0   
1                                     0                                     0   
2                                     0                                     0   
3                                     1                                     0   
4                                     0                                     0   

   Specialization_Retail Management  Specialization_Rural and Agribusiness  \
0                                 0                                      0   
1                                 0                                      0   
2                                 0                                      0   
3                                 0                                      0   
4                                 0                                      0   

   Specialization_Services Excellence  Specialization_Supply Chain Management  \
0                                   0                                       0   
1                                   0                                       0   
2                                   0                                       0   
3                                   0                                       0   
4                                   0                                       0   

   Specialization_Travel and Tourism  
0                                  0  
1                                  0  
2                                  0  
3                                  0  
4                                  0  

[5 rows x 75 columns]

In [32]:
# Split the dataset into Train and Test sets
# Import the required library

from sklearn.model_selection import train_test_split

In [33]:
# Storing Feature variables in X

X = leads.drop(['Converted'], 1)
X.head()

TotalVisits  Total Time Spent on Website  Page Views Per Visit  \
0          0.0                            0                   0.0   
1          5.0                          674                   2.5   
2          2.0                         1532                   2.0   
3          1.0                          305                   1.0   
4          2.0                         1428                   1.0   

   Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
0                                    0                          0   
1                                    0                          0   
2                                    1                          0   
3                                    1                          0   
4                                    1                          0   

   Lead Origin_Lead Import  Lead Source_Direct Traffic  Lead Source_Facebook  \
0                        0                           0                     0   
1                        0                           0                     0   
2                        0                           1                     0   
3                        0                           1                     0   
4                        0                           0                     0   

   Lead Source_Google  Lead Source_Live Chat  ...  \
0                   0                      0  ...   
1                   0                      0  ...   
2                   0                      0  ...   
3                   0                      0  ...   
4                   1                      0  ...   

   Specialization_IT Projects Management  \
0                                      0   
1                                      0   
2                                      0   
3                                      0   
4                                      0   

   Specialization_International Business  Specialization_Marketing Management  \
0                                      0                                    0   
1                                      0                                    0   
2                                      0                                    0   
3                                      0                                    0   
4                                      0                                    0   

   Specialization_Media and Advertising  Specialization_Operations Management  \
0                                     0                                     0   
1                                     0                                     0   
2                                     0                                     0   
3                                     1                                     0   
4                                     0                                     0   

   Specialization_Retail Management  Specialization_Rural and Agribusiness  \
0                                 0                                      0   
1                                 0                                      0   
2                                 0                                      0   
3                                 0                                      0   
4                                 0                                      0   

   Specialization_Services Excellence  Specialization_Supply Chain Management  \
0                                   0                                       0   
1                                   0                                       0   
2                                   0                                       0   
3                                   0                                       0   
4                                   0                                       0   

   Specialization_Travel and Tourism  
0                                  0  
1                                  0  
2                                  0  
3                     

In [34]:
# Storing Target variables in y
y = leads['Converted']
y.head()

0    0
1    0
2    1
3    0
4    1
Name: Converted, dtype: int64

In [35]:
# Splitting the dataset into 70% Train and 30% Test

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=100)

In [36]:
# Rescaling the numeric variables present in the dataset.

# Import MinMax scaler
from sklearn.preprocessing import MinMaxScaler

In [37]:
# Scaling all the numeric variables in x_Train using MinMaxScalar

scaler = MinMaxScaler()

X_train[['TotalVisits', 'Page Views Per Visit', 'Total Time Spent on Website']] = scaler.fit_transform(X_train[['TotalVisits', 'Page Views Per Visit', 'Total Time Spent on Website']])

X_train.head()

TotalVisits  Total Time Spent on Website  Page Views Per Visit  \
8003     0.015936                     0.029489                 0.125   
218      0.015936                     0.082306                 0.250   
4171     0.023904                     0.034331                 0.375   
4037     0.000000                     0.000000                 0.000   
3660     0.000000                     0.000000                 0.000   

      Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
8003                                    1                          0   
218                                     1                          0   
4171                                    1                          0   
4037                                    0                          0   
3660                                    0                          1   

      Lead Origin_Lead Import  Lead Source_Direct Traffic  \
8003                        0                           1   
218                         0                           1   
4171                        0                           1   
4037                        0                           0   
3660                        0                           0   

      Lead Source_Facebook  Lead Source_Google  Lead Source_Live Chat  ...  \
8003                     0                   0                      0  ...   
218                      0                   0                      0  ...   
4171                     0                   0                      0  ...   
4037                     0                   0                      0  ...   
3660                     0                   0                      0  ...   

      Specialization_IT Projects Management  \
8003                                      1   
218                                       0   
4171                                      0   
4037                                      0   
3660                                      0   

      Specialization_International Business  \
8003                                      0   
218                                       0   
4171                                      0   
4037                                      0   
3660                                      0   

      Specialization_Marketing Management  \
8003                                    0   
218                                     0   
4171                                    0   
4037                                    0   
3660                                    0   

      Specialization_Media and Advertising  \
8003                                     0   
218                                      0   
4171                                     0   
4037                                     0   
3660                                     0   

      Specialization_Operations Management  Specialization_Retail Management  \
8003                                     0                                 0   
218                                      0                                 0   
4171                                     0                                 0   
4037                                     0                                 0   
3660                                     0                                 0   

      Specialization_Rural and Agribusiness  \
8003                                      0   
218                                       0   
4171                                      0   
4037                                      0   
3660                                      0   

      Specialization_Services Excellence  \
8003                                   0   
218                                    0   
4171                                   0   
4037                                   0   
3660                                   0   

      Specialization_Supply Chain Management  \
8003                                       0   
218                                        0   
4171             

In [38]:
# Corelation between different variables
leads.corr()

Converted  TotalVisits  \
Converted                                            1.000000     0.005651   
TotalVisits                                          0.005651     1.000000   
Total Time Spent on Website                          0.313338     0.202551   
Page Views Per Visit                                -0.063362     0.489039   
Lead Origin_Landing Page Submission                 -0.117563     0.267954   
Lead Origin_Lead Add Form                            0.288666    -0.208375   
Lead Origin_Lead Import                             -0.019269    -0.043000   
Lead Source_Direct Traffic                          -0.133600     0.075252   
Lead Source_Facebook                                -0.021207    -0.042052   
Lead Source_Google                                   0.020205     0.085306   
Lead Source_Live Chat                                0.018407    -0.012720   
Lead Source_Olark Chat                              -0.044348    -0.247963   
Lead Source_Organic Search                          -0.034223     0.192469   
Lead Source_Pay per Click Ads                       -0.012059    -0.001549   
Lead Source_Press_Release                           -0.012059     0.005896   
Lead Source_Reference                                0.243200    -0.180458   
Lead Source_Referral Sites                          -0.023501     0.078163   
Lead Source_Social Media                             0.000676    -0.003946   
Lead Source_WeLearn                                  0.013015    -0.004031   
Lead Source_Welingak Website                         0.144864    -0.097011   
Lead Source_bing                                    -0.006411    -0.004116   
Lead Source_testone                                 -0.012059     0.003414   
Do Not Email_Yes                                    -0.148451     0.028900   
Last Activity_Converted to Lead                     -0.138834    -0.074517   
Last Activity_Email Bounced                         -0.134838    -0.040748   
Last Activity_Email Link Clicked                    -0.033557    -0.013791   
Last Activity_Email Marked Spam                      0.018407     0.001319   
Last Activity_Email Opened                          -0.022391     0.011982   
Last Activity_Email Received                         0.018407     0.006584   
Last Activity_Form Submitted on Website             -0.036322     0.007614   
...                                                       ...          ...   
Last Notable Activity_Email Link Clicked            -0.026291     0.002706   
Last Notable Activity_Email Marked Spam              0.018407     0.001319   
Last Notable Activity_Email Opened                  -0.010199     0.026851   
Last Notable Activity_Email Received                 0.013015     0.005896   
Last Notable Activity_Had a Phone Conversation       0.046969     0.026100   
Last Notable Activity_Modified                      -0.239572    -0.088690   
Last Notable Activity_Olark Chat Conversation       -0.039205    -0.004994   
Last Notable Activity_Page Visited on Website       -0.032219     0.226728   
Last Notable Activity_SMS Sent                       0.284301    -0.028923   
Last Notable Activity_Unreachable                    0.050436     0.002792   
Last Notable Activity_Unsubscribed                  -0.026673     0.001631   
Last Notable Activity_View in browser link Clicked  -0.012059     0.010859   
Specialization_Banking, Investment And Insurance     0.036229     0.042269   
Specialization_Business Administration               0.014468     0.035184   
Specialization_E-Business                           -0.017955     0.015248   
Specialization_E-COMMERCE                           -0.015443     0.019003   
Specialization_Finance Management                    0.027078     0.031922   
Specialization_Healthcare Management                 0.028249     0.027640   
Specialization_Hospitality Management               -0.006080     0.016549   
Specialization_Human Resource Management             0.025865     0.032934

In [39]:
# # For Visualisation
# import matplotlib.pyplot as plt
# import seaborn as sns
# %matplotlib inline
# plt.figure(figsize = (20,10))        
# sns.heatmap(leads.corr(),annot = True)

Building Models using RFE

In [40]:
# Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# Creating a LogisticRegression object log_reg
log_reg = LogisticRegression()

In [41]:
# Import RFE
from sklearn.feature_selection import RFE

# Select 20 variables
rfe = RFE(log_reg, 20)             
rfe = rfe.fit(X_train, y_train)

In [42]:
# Let's take a look at which features have been selected by RFE

list(zip(X_train.columns, rfe.support_, rfe.ranking_))

[('TotalVisits', True, 1),
 ('Total Time Spent on Website', True, 1),
 ('Page Views Per Visit', False, 2),
 ('Lead Origin_Landing Page Submission', False, 6),
 ('Lead Origin_Lead Add Form', True, 1),
 ('Lead Origin_Lead Import', False, 51),
 ('Lead Source_Direct Traffic', False, 18),
 ('Lead Source_Facebook', False, 46),
 ('Lead Source_Google', False, 31),
 ('Lead Source_Live Chat', False, 37),
 ('Lead Source_Olark Chat', True, 1),
 ('Lead Source_Organic Search', False, 30),
 ('Lead Source_Pay per Click Ads', False, 36),
 ('Lead Source_Press_Release', False, 47),
 ('Lead Source_Reference', True, 1),
 ('Lead Source_Referral Sites', False, 32),
 ('Lead Source_Social Media', False, 52),
 ('Lead Source_WeLearn', False, 33),
 ('Lead Source_Welingak Website', True, 1),
 ('Lead Source_bing', False, 27),
 ('Lead Source_testone', False, 29),
 ('Do Not Email_Yes', True, 1),
 ('Last Activity_Converted to Lead', False, 3),
 ('Last Activity_Email Bounced', True, 1),
 ('Last Activity_Email Link Clic

In [43]:
# Creating a variable rfe_col and storing all the columns selected by RFE

rfe_col = X_train.columns[rfe.support_]

In [44]:
X_train = X_train[rfe_col]

In [45]:
# Import statsmodels

import statsmodels.api as sm

In [46]:
# Fit a logistic Regression model on X_train after adding a constant and output the summary

X_train_sm = sm.add_constant(X_train)
log_mod = sm.GLM(y_train, X_train_sm, family = sm.families.Binomial())
reg_res = log_mod.fit()
reg_res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4440
Model Family:                Binomial   Df Model:                           20
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2019.4
Date:                Fri, 15 Nov 2019   Deviance:                       4038.7
Time:                        16:09:24   Pearson chi2:                 4.69e+03
No. Iterations:                    22   Covariance Type:             nonrobust
========================================================================================================================
                                                           coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------
const                                                   -0.7083      0.607     -1.166      0.244      -1.899       0.482
TotalVisits                                              8.6822      2.585      3.359      0.001       3.615      13.749
Total Time Spent on Website                              4.4003      0.187     23.550      0.000       4.034       4.766
Lead Origin_Lead Add Form                                2.9442      1.210      2.433      0.015       0.572       5.316
Lead Source_Olark Chat                                   1.5531      0.127     12.219      0.000       1.304       1.802
Lead Source_Reference                                    1.2610      1.233      1.023      0.306      -1.156       3.678
Lead Source_Welingak Website                             3.3115      1.573      2.106      0.035       0.229       6.394
Do Not Email_Yes                                        -1.4072      0.223     -6.310      0.000      -1.844      -0.970
Last Activity_Email Bounced                             -1.0803      0.662     -1.631      0.103      -2.378       0.218
Last Activity_Had a Phone Conversation                   1.4850      0.984      1.510      0.131      -0.443       3.413
Last Activity_Olark Chat Conversation                   -0.6503      0.192     -3.393      0.001      -1.026      -0.275
Last Activity_SMS Sent                                   0.9648      0.086     11.232      0.000       0.796       1.133
What is your current occupation_Housewife               22.5870    2.4e+04      0.001      0.999    -4.7e+04     4.7e+04
What is your current occupation_Student                 -1.1422      0.640     -1.785      0.074      -2.396       0.112
What is your current occupation_Unemployed              -1.3134      0.601     -2.184      0.029      -2.492      -0.135
What is your current occupation_Working Professional     1.2988      0.631      2.058      0.040       0.062       2.536
Last Notable Activity_Email Bounced                      1.4182      0.798      1.778      0.075      -0.145       2.982
Last Notable Activity_Had a Phone Conversation          22.3445   2.12e+04      0.001      0.999   -4.15e+04    4.16e+04
Last Notable Activity_Modified                          -0.7018      0.095     -7.414      0.000      -0.887      -0.516
Last Notable Activity_Unreachable                        2.5031      0.808      3.099      0.002       0.920       4.086
Specialization_Banking, Investment And Insurance         0.5498      0.204      2.692      0.007       0.150       0.950
========================================================================================================================
"""

In [47]:
# Import 'variance_inflation_factor'

from statsmodels.stats.outliers_influence import variance_inflation_factor

In [48]:
# Make a VIF dataframe for all the variables present

vif = pd.DataFrame()
vif['Features'] = X_train.columns
vif['VIF'] = [variance_inflation_factor(X_train.values, i) for i in range(X_train.shape[1])]
vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values(by = "VIF", ascending = False)
vif

Features    VIF
2                           Lead Origin_Lead Add Form  84.21
4                               Lead Source_Reference  65.21
5                        Lead Source_Welingak Website  20.03
13         What is your current occupation_Unemployed   4.63
8              Last Activity_Had a Phone Conversation   2.45
16     Last Notable Activity_Had a Phone Conversation   2.45
1                         Total Time Spent on Website   2.40
7                         Last Activity_Email Bounced   2.01
17                     Last Notable Activity_Modified   1.77
10                             Last Activity_SMS Sent   1.74
6                                    Do Not Email_Yes   1.65
14  What is your current occupation_Working Profes...   1.65
0                                         TotalVisits   1.64
3                              Lead Source_Olark Chat   1.54
15                Last Notable Activity_Email Bounced   1.41
9               Last Activity_Olark Chat Conversation   1.30
12            What is your current occupation_Student   1.12
19   Specialization_Banking, Investment And Insurance   1.05
11          What is your current occupation_Housewife   1.01
18                  Last Notable Activity_Unreachable   1.01

Lets drop the variable Lead Source_Reference as the VIF value is high i.e 65.21 and p-value is also high with 0.306

In [49]:
X_train.drop('Lead Source_Reference', axis = 1, inplace = True)

# Refit the model
log_mod1 = sm.GLM(y_train,(sm.add_constant(X_train)), family = sm.families.Binomial())
log_mod1.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4441
Model Family:                Binomial   Df Model:                           19
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2019.8
Date:                Fri, 15 Nov 2019   Deviance:                       4039.6
Time:                        16:09:25   Pearson chi2:                 4.67e+03
No. Iterations:                    22   Covariance Type:             nonrobust
========================================================================================================================
                                                           coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------
const                                                   -0.7076      0.607     -1.165      0.244      -1.898       0.483
TotalVisits                                              8.6793      2.585      3.357      0.001       3.612      13.746
Total Time Spent on Website                              4.4003      0.187     23.549      0.000       4.034       4.766
Lead Origin_Lead Add Form                                4.1668      0.260     16.019      0.000       3.657       4.677
Lead Source_Olark Chat                                   1.5529      0.127     12.217      0.000       1.304       1.802
Lead Source_Welingak Website                             2.0887      1.037      2.014      0.044       0.056       4.121
Do Not Email_Yes                                        -1.4049      0.223     -6.306      0.000      -1.842      -0.968
Last Activity_Email Bounced                             -1.0819      0.662     -1.634      0.102      -2.380       0.216
Last Activity_Had a Phone Conversation                   1.4853      0.984      1.510      0.131      -0.443       3.413
Last Activity_Olark Chat Conversation                   -0.6498      0.192     -3.391      0.001      -1.025      -0.274
Last Activity_SMS Sent                                   0.9649      0.086     11.234      0.000       0.797       1.133
What is your current occupation_Housewife               22.5884    2.4e+04      0.001      0.999    -4.7e+04    4.71e+04
What is your current occupation_Student                 -1.1419      0.640     -1.785      0.074      -2.396       0.112
What is your current occupation_Unemployed              -1.3143      0.601     -2.186      0.029      -2.493      -0.136
What is your current occupation_Working Professional     1.2990      0.631      2.059      0.039       0.062       2.536
Last Notable Activity_Email Bounced                      1.4176      0.798      1.777      0.076      -0.146       2.981
Last Notable Activity_Had a Phone Conversation          22.3444   2.12e+04      0.001      0.999   -4.15e+04    4.16e+04
Last Notable Activity_Modified                          -0.7020      0.095     -7.417      0.000      -0.887      -0.516
Last Notable Activity_Unreachable                        2.5036      0.808      3.099      0.002       0.920       4.087
Specialization_Banking, Investment And Insurance         0.5506      0.204      2.697      0.007       0.150       0.951
========================================================================================================================
"""

In [50]:
vif = pd.DataFrame()
vif['Features'] = X_train.columns
vif['VIF'] = [variance_inflation_factor(X_train.values, i) for i in range(X_train.shape[1])]
vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values(by = "VIF", ascending = False)
vif

Features   VIF
12         What is your current occupation_Unemployed  4.63
15     Last Notable Activity_Had a Phone Conversation  2.45
7              Last Activity_Had a Phone Conversation  2.45
1                         Total Time Spent on Website  2.39
6                         Last Activity_Email Bounced  2.01
16                     Last Notable Activity_Modified  1.77
9                              Last Activity_SMS Sent  1.74
2                           Lead Origin_Lead Add Form  1.71
5                                    Do Not Email_Yes  1.65
13  What is your current occupation_Working Profes...  1.65
0                                         TotalVisits  1.64
3                              Lead Source_Olark Chat  1.54
14                Last Notable Activity_Email Bounced  1.41
4                        Lead Source_Welingak Website  1.33
8               Last Activity_Olark Chat Conversation  1.30
11            What is your current occupation_Student  1.12
18   Specialization_Banking, Investment And Insurance  1.05
10          What is your current occupation_Housewife  1.01
17                  Last Notable Activity_Unreachable  1.01

In [51]:
# Dropping Last Notable Activity_Had a Phone Conversation as p value is 0.999
X_train.drop('Last Notable Activity_Had a Phone Conversation', axis = 1, inplace = True)

In [52]:
# Refit the model
log_mod2 = sm.GLM(y_train,(sm.add_constant(X_train)), family = sm.families.Binomial())
log_mod2.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4442
Model Family:                Binomial   Df Model:                           18
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2021.6
Date:                Fri, 15 Nov 2019   Deviance:                       4043.2
Time:                        16:09:25   Pearson chi2:                 4.67e+03
No. Iterations:                    21   Covariance Type:             nonrobust
========================================================================================================================
                                                           coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------
const                                                   -0.7046      0.607     -1.160      0.246      -1.895       0.486
TotalVisits                                              8.7256      2.586      3.374      0.001       3.658      13.794
Total Time Spent on Website                              4.3956      0.187     23.529      0.000       4.029       4.762
Lead Origin_Lead Add Form                                4.1678      0.260     16.019      0.000       3.658       4.678
Lead Source_Olark Chat                                   1.5524      0.127     12.212      0.000       1.303       1.802
Lead Source_Welingak Website                             2.0879      1.037      2.013      0.044       0.055       4.121
Do Not Email_Yes                                        -1.4054      0.223     -6.309      0.000      -1.842      -0.969
Last Activity_Email Bounced                             -1.0741      0.662     -1.622      0.105      -2.372       0.224
Last Activity_Had a Phone Conversation                   2.6597      0.801      3.320      0.001       1.089       4.230
Last Activity_Olark Chat Conversation                   -0.6449      0.192     -3.365      0.001      -1.020      -0.269
Last Activity_SMS Sent                                   0.9636      0.086     11.220      0.000       0.795       1.132
What is your current occupation_Housewife               21.5890   1.46e+04      0.001      0.999   -2.85e+04    2.86e+04
What is your current occupation_Student                 -1.1433      0.640     -1.787      0.074      -2.397       0.111
What is your current occupation_Unemployed              -1.3146      0.601     -2.186      0.029      -2.493      -0.136
What is your current occupation_Working Professional     1.2978      0.631      2.057      0.040       0.061       2.534
Last Notable Activity_Email Bounced                      1.4085      0.798      1.766      0.077      -0.155       2.972
Last Notable Activity_Modified                          -0.7111      0.095     -7.523      0.000      -0.896      -0.526
Last Notable Activity_Unreachable                        2.5009      0.808      3.096      0.002       0.918       4.084
Specialization_Banking, Investment And Insurance         0.5644      0.203      2.776      0.005       0.166       0.963
========================================================================================================================
"""

In [53]:
# Dropping What is your current occupation_Housewife feature
X_train.drop('What is your current occupation_Housewife', axis = 1, inplace = True)

In [54]:
# Refit the model
log_mod3 = sm.GLM(y_train,(sm.add_constant(X_train)), family = sm.families.Binomial())
log_mod3.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4443
Model Family:                Binomial   Df Model:                           17
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2023.9
Date:                Fri, 15 Nov 2019   Deviance:                       4047.7
Time:                        16:09:25   Pearson chi2:                 4.68e+03
No. Iterations:                     7   Covariance Type:             nonrobust
========================================================================================================================
                                                           coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------
const                                                   -0.1300      0.558     -0.233      0.816      -1.224       0.964
TotalVisits                                              8.5476      2.574      3.321      0.001       3.503      13.592
Total Time Spent on Website                              4.3937      0.187     23.527      0.000       4.028       4.760
Lead Origin_Lead Add Form                                4.1644      0.260     16.013      0.000       3.655       4.674
Lead Source_Olark Chat                                   1.5476      0.127     12.185      0.000       1.299       1.797
Lead Source_Welingak Website                             2.0878      1.037      2.013      0.044       0.055       4.121
Do Not Email_Yes                                        -1.4095      0.223     -6.317      0.000      -1.847      -0.972
Last Activity_Email Bounced                             -1.0726      0.662     -1.620      0.105      -2.370       0.225
Last Activity_Had a Phone Conversation                   2.6604      0.801      3.321      0.001       1.090       4.230
Last Activity_Olark Chat Conversation                   -0.6447      0.192     -3.365      0.001      -1.020      -0.269
Last Activity_SMS Sent                                   0.9635      0.086     11.220      0.000       0.795       1.132
What is your current occupation_Student                 -1.7140      0.595     -2.880      0.004      -2.880      -0.548
What is your current occupation_Unemployed              -1.8852      0.554     -3.405      0.001      -2.970      -0.800
What is your current occupation_Working Professional     0.7269      0.585      1.242      0.214      -0.420       1.874
Last Notable Activity_Email Bounced                      1.4114      0.798      1.770      0.077      -0.152       2.975
Last Notable Activity_Modified                          -0.7104      0.094     -7.523      0.000      -0.896      -0.525
Last Notable Activity_Unreachable                        2.4982      0.808      3.092      0.002       0.914       4.082
Specialization_Banking, Investment And Insurance         0.5645      0.203      2.777      0.005       0.166       0.963
========================================================================================================================
"""

In [55]:
vif = pd.DataFrame()
vif['Features'] = X_train.columns
vif['VIF'] = [variance_inflation_factor(X_train.values, i) for i in range(X_train.shape[1])]
vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values(by = "VIF", ascending = False)
vif

Features   VIF
11         What is your current occupation_Unemployed  4.61
1                         Total Time Spent on Website  2.38
6                         Last Activity_Email Bounced  2.01
14                     Last Notable Activity_Modified  1.75
9                              Last Activity_SMS Sent  1.74
2                           Lead Origin_Lead Add Form  1.70
5                                    Do Not Email_Yes  1.65
0                                         TotalVisits  1.64
12  What is your current occupation_Working Profes...  1.64
3                              Lead Source_Olark Chat  1.53
13                Last Notable Activity_Email Bounced  1.41
4                        Lead Source_Welingak Website  1.33
8               Last Activity_Olark Chat Conversation  1.30
10            What is your current occupation_Student  1.12
16   Specialization_Banking, Investment And Insurance  1.05
7              Last Activity_Had a Phone Conversation  1.01
15                  Last Notable Activity_Unreachable  1.01

In [56]:
# Dropping Last Activity_Email Bounced feature
X_train.drop('Last Activity_Email Bounced', axis = 1, inplace = True)

In [57]:
# Refit the model
log_mod4 = sm.GLM(y_train,(sm.add_constant(X_train)), family = sm.families.Binomial())
log_mod4.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4444
Model Family:                Binomial   Df Model:                           16
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2025.5
Date:                Fri, 15 Nov 2019   Deviance:                       4051.0
Time:                        16:09:25   Pearson chi2:                 4.66e+03
No. Iterations:                     7   Covariance Type:             nonrobust
========================================================================================================================
                                                           coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------
const                                                   -0.1211      0.558     -0.217      0.828      -1.216       0.973
TotalVisits                                              8.8196      2.572      3.429      0.001       3.778      13.861
Total Time Spent on Website                              4.3927      0.187     23.534      0.000       4.027       4.758
Lead Origin_Lead Add Form                                4.1835      0.260     16.074      0.000       3.673       4.694
Lead Source_Olark Chat                                   1.5462      0.127     12.180      0.000       1.297       1.795
Lead Source_Welingak Website                             2.0933      1.038      2.018      0.044       0.060       4.127
Do Not Email_Yes                                        -1.5399      0.211     -7.295      0.000      -1.954      -1.126
Last Activity_Had a Phone Conversation                   2.6677      0.801      3.328      0.001       1.097       4.239
Last Activity_Olark Chat Conversation                   -0.6246      0.191     -3.264      0.001      -1.000      -0.250
Last Activity_SMS Sent                                   0.9755      0.086     11.383      0.000       0.808       1.143
What is your current occupation_Student                 -1.7276      0.595     -2.903      0.004      -2.894      -0.561
What is your current occupation_Unemployed              -1.8987      0.554     -3.429      0.001      -2.984      -0.814
What is your current occupation_Working Professional     0.7060      0.585      1.207      0.228      -0.441       1.853
Last Notable Activity_Email Bounced                      0.4708      0.520      0.906      0.365      -0.548       1.489
Last Notable Activity_Modified                          -0.7317      0.094     -7.798      0.000      -0.916      -0.548
Last Notable Activity_Unreachable                        2.4987      0.808      3.092      0.002       0.915       4.082
Specialization_Banking, Investment And Insurance         0.5642      0.203      2.774      0.006       0.166       0.963
========================================================================================================================
"""

In [58]:
# Dropping What is your current occupation_Working Professional feature
X_train.drop('What is your current occupation_Working Professional', axis = 1, inplace = True)

In [59]:
# Refit the model
log_mod5 = sm.GLM(y_train,(sm.add_constant(X_train)), family = sm.families.Binomial())
log_mod5.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4445
Model Family:                Binomial   Df Model:                           15
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2026.2
Date:                Fri, 15 Nov 2019   Deviance:                       4052.3
Time:                        16:09:25   Pearson chi2:                 4.63e+03
No. Iterations:                     7   Covariance Type:             nonrobust
====================================================================================================================
                                                       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------------------
const                                                0.5173      0.201      2.575      0.010       0.124       0.911
TotalVisits                                          8.7289      2.567      3.400      0.001       3.697      13.761
Total Time Spent on Website                          4.3928      0.187     23.532      0.000       4.027       4.759
Lead Origin_Lead Add Form                            4.1835      0.260     16.077      0.000       3.673       4.694
Lead Source_Olark Chat                               1.5457      0.127     12.181      0.000       1.297       1.794
Lead Source_Welingak Website                         2.0916      1.038      2.016      0.044       0.058       4.125
Do Not Email_Yes                                    -1.5394      0.211     -7.291      0.000      -1.953      -1.126
Last Activity_Had a Phone Conversation               2.6716      0.801      3.334      0.001       1.101       4.242
Last Activity_Olark Chat Conversation               -0.6206      0.191     -3.246      0.001      -0.995      -0.246
Last Activity_SMS Sent                               0.9790      0.086     11.439      0.000       0.811       1.147
What is your current occupation_Student             -2.3658      0.289     -8.198      0.000      -2.931      -1.800
What is your current occupation_Unemployed          -2.5369      0.188    -13.468      0.000      -2.906      -2.168
Last Notable Activity_Email Bounced                  0.4783      0.519      0.922      0.357      -0.539       1.495
Last Notable Activity_Modified                      -0.7328      0.094     -7.810      0.000      -0.917      -0.549
Last Notable Activity_Unreachable                    2.5000      0.808      3.094      0.002       0.916       4.084
Specialization_Banking, Investment And Insurance     0.5659      0.203      2.784      0.005       0.168       0.964
====================================================================================================================
"""

In [60]:
# Dropping Last Notable Activity_Email Bounced feature
X_train.drop('Last Notable Activity_Email Bounced', axis = 1, inplace = True)

In [61]:
# Refit the model
log_mod6 = sm.GLM(y_train,(sm.add_constant(X_train)), family = sm.families.Binomial())
log_mod6.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 4461
Model:                            GLM   Df Residuals:                     4446
Model Family:                Binomial   Df Model:                           14
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2026.6
Date:                Fri, 15 Nov 2019   Deviance:                       4053.1
Time:                        16:09:25   Pearson chi2:                 4.63e+03
No. Iterations:                     7   Covariance Type:             nonrobust
====================================================================================================================
                                                       coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------------------
const                                                0.5243      0.201      2.609      0.009       0.130       0.918
TotalVisits                                          8.6504      2.570      3.366      0.001       3.613      13.688
Total Time Spent on Website                          4.3937      0.187     23.540      0.000       4.028       4.759
Lead Origin_Lead Add Form                            4.1775      0.260     16.068      0.000       3.668       4.687
Lead Source_Olark Chat                               1.5459      0.127     12.183      0.000       1.297       1.795
Lead Source_Welingak Website                         2.0860      1.037      2.011      0.044       0.053       4.119
Do Not Email_Yes                                    -1.4684      0.194     -7.567      0.000      -1.849      -1.088
Last Activity_Had a Phone Conversation               2.6713      0.801      3.334      0.001       1.101       4.242
Last Activity_Olark Chat Conversation               -0.6211      0.191     -3.249      0.001      -0.996      -0.246
Last Activity_SMS Sent                               0.9716      0.085     11.412      0.000       0.805       1.138
What is your current occupation_Student             -2.3685      0.289     -8.203      0.000      -2.934      -1.803
What is your current occupation_Unemployed          -2.5388      0.189    -13.464      0.000      -2.908      -2.169
Last Notable Activity_Modified                      -0.7384      0.094     -7.890      0.000      -0.922      -0.555
Last Notable Activity_Unreachable                    2.4960      0.808      3.089      0.002       0.912       4.080
Specialization_Banking, Investment And Insurance     0.5640      0.203      2.778      0.005       0.166       0.962
====================================================================================================================
"""

In [62]:
vif = pd.DataFrame()
vif['Features'] = X_train.columns
vif['VIF'] = [variance_inflation_factor(X_train.values, i) for i in range(X_train.shape[1])]
vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values(by = "VIF", ascending = False)
vif

Features   VIF
10        What is your current occupation_Unemployed  3.37
1                        Total Time Spent on Website  2.00
11                    Last Notable Activity_Modified  1.61
8                             Last Activity_SMS Sent  1.59
0                                        TotalVisits  1.54
2                          Lead Origin_Lead Add Form  1.46
3                             Lead Source_Olark Chat  1.43
4                       Lead Source_Welingak Website  1.31
7              Last Activity_Olark Chat Conversation  1.29
5                                   Do Not Email_Yes  1.09
9            What is your current occupation_Student  1.09
13  Specialization_Banking, Investment And Insurance  1.05
6             Last Activity_Had a Phone Conversation  1.01
12                 Last Notable Activity_Unreachable  1.01